EIPS/eip-9999.md

```diff
@@ -0,0 +1,150 @@
+     * If no:
+       * Charge `GAS_COLD_SLOAD`
+       * Add storage key to `block_level_accessed_storage_keys`
+    
+
+### Implementation Details
+
+The implementation modifies the block execution process to maintain block-level sets of accessed addresses and storage slots. 
+
+#### Block-Level Storage Management
+
+```python
+def apply_body(...):
+    # Initialize block-level tracking sets
+    block_level_accessed_addresses = set()
+    block_level_accessed_storage_keys = set()
+    
+    for i, tx in enumerate(map(decode_transaction, transactions)):
+        # Create environment with block-level context
+        env = vm.Environment(
+            # ... other parameters ...
+            block_level_accessed_addresses=block_level_accessed_addresses,
+            block_level_accessed_storage_keys=block_level_accessed_storage_keys
+        )
+        
+        # Process transaction and update block-level sets
+        gas_used, accessed_addresses, accessed_storage_keys, logs, error = process_transaction(env, tx)
+        block_level_accessed_addresses += accessed_addresses
+        block_level_accessed_storage_keys += accessed_storage_keys
+```
+
+This code shows how block-level slot warming is implemented at the execution layer. The `block_level_accessed_addresses` and `block_level_accessed_storage_keys` sets are maintained throughout the block's execution and passed to each transaction's environment.
+
+#### Transaction Processing
+
+```python
+def process_transaction(env: vm.Environment, tx: Transaction) -> Tuple[Uint, Tuple[Log, ...], Optional[Exception]]:
+    preaccessed_addresses = set()
+    preaccessed_storage_keys = set()
+    
+    # Add block-level pre-accessed slots
+    preaccessed_addresses.add(env.block_level_accessed_addresses)
+    preaccessed_storage_keys.add(env.block_level_accessed_storage_keys)
+    
+    # Handle access lists from transaction
+    ...
+```
+
+This adds the block-level-accessed addresses and storage keys to the preaccessed addresses and storage keys.
+As a result, from the perspective of a transaction, block-level accessed addresses and storage keys are treated the same as precompiles or the coinbase address.
+
+```python
+def process_message_call(message: Message, env: Environment) -> MessageCallOutput:
+    return MessageCallOutput(
+        # ... other fields ...
+        accessed_addresses=evm.accessed_addresses,
+        accessed_storage_keys=evm.accessed_storage_keys
+    )
+```
+
+The message call processing tracks accessed addresses and storage keys during execution, which are then propagated back up to the transaction level and ultimately to the block level.
+
+## Rationale
+
+The proposal builds on several key observations:
+
+1. **Caching Efficiency**: Modern Ethereum clients already implement sophisticated caching mechanisms at the block level. Extending address and storage key warming to match this caching behavior better aligns gas costs with actual computational costs.
```

Comment:
Right, this caching point would also show why I think we need to remove the EIP-2929 reverting mechanism. The "cold" gas cost represents the expensive price we have to pay if we have to read the storage slot from disk. Even if a sub-call reverts, then most (all (?)) clients should have kept this in memory and should not read this again from disk. Therefore, only the first access should charge the cold cost, irrelevant if it invalidates this sub-call / transaction later.

Reply:
Put it into the rational section for now but might be worth to directly adopt "reverts don't unwarm slots" in the EIP